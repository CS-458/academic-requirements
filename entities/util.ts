let id = 0;
export function uniqueId(): string {
  return `id_${id++}`;
}
