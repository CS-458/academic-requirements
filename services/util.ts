/// Fetch a request (similar to fetch), but throw an error if the request fails
export async function fetchApi<T>(url: string, init?: RequestInit): Promise<T> {
  const res = await fetch(url, init);
  if (!res.ok) {
    throw new Error(`Response code ${res.status} for request to ${url}`);
  }
  return await res.json();
}
