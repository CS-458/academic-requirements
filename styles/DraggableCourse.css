.CourseText {
  font-size: 9pt;
  text-align: center;
  overflow: hidden;
  text-overflow: ellipsis;
  width: 100%;
  /*display: -webkit-box;*/
  -webkit-line-clamp: 3;
  -webkit-box-orient: horizontal;
  border-width: 2px;
  border-color: black;
  border-style: solid;
  border-radius: 5px;
  padding: .5em 1em;
  margin: 5px 0px;
  color: black;
  line-height: 1.1em;
}

.CourseText {
  background-color: gainsboro;
}

.CourseWarningYellow {
  background-color: rgb(219, 248, 90);
}

.CourseWarningOrange {
  background-color: rgb(231, 159, 77);
}

.CourseWarningRed {
  background-color: rgb(214, 53, 53);
}

.CourseText:first-line {
  font-size: 11pt;
}
