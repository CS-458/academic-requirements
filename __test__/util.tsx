import { jest } from "@jest/globals";
import type { MatcherFunction } from "expect";
import { DndProvider } from "react-dnd";
import { HTML5Backend } from "react-dnd-html5-backend";
import { QueryClient, QueryClientProvider, UseQueryResult } from "react-query";
import { testApiHandler } from "next-test-api-route-handler";
import { NextApiHandler } from "next";
import {
  Matcher,
  screen,
  render as testRender,
  RenderResult,
  waitFor
} from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { UserEvent } from "@testing-library/user-event/dist/types/setup/setup";
import { User, UserMajor } from "../services/user";
import { VerifyIdTokenOptions, LoginTicket } from "google-auth-library";
import { updateClient } from "../services/login";
import { PromisedDatabase } from "promised-sqlite3";
import { setUserDb } from "../services/sql";

import "../pages/api";
import { dragAndDrop } from "./dragDrop";

interface UserExt {
  /// Select an element from an `AutoComplete` dropdown
  selectAutocomplete: typeof selectAutocomplete;
  /// Drag an element from one place to another
  drag: typeof dragAndDrop;
}

async function selectAutocomplete(
  this: UserEvent,
  label: Matcher,
  option: Matcher
): Promise<void> {
  await this.click(screen.getByTestId("test-root-element"));
  const input = screen.getByLabelText(label, { selector: "input" });
  await waitFor(async () => {
    expect(input).not.toBeDisabled();
  });
  await this.click(input);
  try {
    await this.click(
      await screen.findByText(option, undefined, { timeout: 10000 })
    );
  } catch (e) {
    const rootEl = screen.getByTestId("test-root-element");
    const popper = rootEl?.querySelector(".MuiAutocomplete-popper");
    const html = popper?.innerHTML;
    throw new Error(
      `Option ${option} was not found in dropdown.\nOptions: ${html}`
    );
  }
}

export function parentEl(el: HTMLElement, name: string): HTMLElement {
  while (!el.classList.contains(`${name}-root`)) {
    const next = el.parentElement;
    if (next == null) {
      throw new Error(`Element does not have a wrapper with name ${name}`);
    }
    el = next;
  }
  return el;
}

/// userEvent.setup(), but with some additional convience methods
export function setupUser(): UserEvent & UserExt {
  return {
    ...userEvent.setup(),
    selectAutocomplete,
    drag: dragAndDrop
  };
}

// const modules: { [key: string]: NextApiHandler | Promise<NextApiHandler> } = {};
// declare function fetch2(input: RequestInfo | URL, init?: RequestInit): Promise<Response>;
/// Executes a request against an Api Route, roughly equavelent with `fetch`.
export async function fetchApiRoute(
  url: RequestInfo | URL,
  init?: RequestInit
): Promise<Response> {
  const res = { resolve: (_a: any) => { }, reject: (_e: any) => { } };
  const result: Promise<Response> = new Promise((resolve, reject) => {
    res.resolve = resolve;
    res.reject = reject;
  });

  let fullUrl: string | undefined;
  let path;
  let params;
  if (url instanceof URL) {
    fullUrl = url.href;
    path = url.pathname;
    params = url.search;
  } else if (typeof url === "string") {
    fullUrl = url;
    [path, params] = url.split("?");
  } else {
    fullUrl = url.url;
    [path, params] = url.url.split("?");
  }

  const handler: NextApiHandler = await import(`../pages/${path}`).catch(
    (e) => {
      if (jest.isEnvironmentTornDown()) {
        console.log("Request made after environment torn down");
      } else {
        console.error(e);
      }
    }
  );
  await testApiHandler({
    url: fullUrl,
    handler,
    params: params?.split("&").reduce((a, b) => {
      const [name, value] = b.split("=");
      return { ...a, [name]: value };
    }, {}),
    test: async ({ fetch }) => {
      const inner = await fetch(init);
      res.resolve(inner);
    }
  });
  return await result;
}

/// Calls fetchApiRoute, and decodes the response as JSON
export async function fetchApiJson(
  url: RequestInfo | URL,
  init?: RequestInit
): Promise<any> {
  return await (await fetchApiRoute(url, init)).json();
}

/**
 *  Wrap Elements with QueryClient & DndProvider similar to _app.tsx
 *
 *  Also overides window.fetch to use fetchApiRoute for testing
 * */
export function render(children: JSX.Element | JSX.Element[]): RenderResult {
  window.fetch = fetchApiRoute;
  return testRender(
    <div data-testid="test-root-element">
      <QueryClientProvider client={new QueryClient()}>
        <DndProvider backend={HTML5Backend}>{children}</DndProvider>
      </QueryClientProvider>
    </div>
  );
}

const localStorage: { [key: string]: string | null } = {};

export function getMockStorage(key: string): string | null {
  return localStorage[key];
}

export function setMockStorage(key: string, value: string): void {
  localStorage[key] = value;
}

export function removeMockStorage(key: string): void {
  localStorage[key] = null;
}

export function clearMockStorage(): void {
  for (const k in localStorage) {
    localStorage[k] = null;
  }
}

/// Configure local storage and set pre-existing data in local storage
export function buildLocalStorage(user?: UserMajor): void {
  window.localStorage.getItem = getMockStorage;
  window.localStorage.setItem = setMockStorage;
  window.localStorage.removeItem = removeMockStorage;
  window.localStorage.clear = clearMockStorage;
  if (user !== undefined) {
    window.localStorage.setItem("user_major", JSON.stringify(user));
  }
}

export function mockUseQuery<T>(
  _key: any,
  fetch: () => Promise<T>
): UseQueryResult<T> & { fetch: Promise<T> } {
  return {
    fetch: fetch(),
    data: undefined,
    error: null,
    isError: false,
    isIdle: false,
    isLoading: true,
    isLoadingError: false,
    isRefetchError: false,
    isSuccess: false,
    status: "loading",
    dataUpdatedAt: Date.now(),
    errorUpdatedAt: Date.now(),
    failureCount: 0,
    errorUpdateCount: 0,
    isFetched: true,
    isFetchedAfterMount: true,
    isFetching: false,
    isPlaceholderData: false,
    isPreviousData: false,
    isRefetching: false,
    isStale: false,
    refetch: async (_options: any) => {
      throw new Error("");
    },
    remove: () => { }
  };
}

export function setupStandaloneQuery(): void {
  jest.mock("react-query", () => ({
    useQuery: mockUseQuery
  }));
}

export async function waitForQuery<T>(
  query: () => UseQueryResult<T>
): Promise<T> {
  // @ts-expect-error fetch only exists on our mock
  const temp = await query().fetch;
  return temp;
}

const toBeUnique: MatcherFunction<[]> = (actual) => {
  if (!Array.isArray(actual)) {
    throw new Error("actual is not an object");
  }
  if (actual === null) {
    return {
      message: () => "null is not an array",
      pass: false
    };
  }
  for (const i of actual) {
    const pos = actual.find((val, idx) => idx > i && actual[i] === val);
    if (pos !== undefined) {
      return {
        message: () => `Indicies ${i}, and ${pos} are the same`,
        pass: false
      };
    }
  }
  return {
    message: () => "All elements are unique",
    pass: true
  };
};

const toBeHidden: MatcherFunction<[]> = (actual) => {
  if (!(actual instanceof HTMLElement)) {
    throw new Error("actual is not an HTMLElement");
  }
  let cur = actual;
  console.log("child", cur.ariaHidden);
  while (cur.ariaHidden == null || cur.ariaHidden === "false") {
    if (cur.dataset.testid === "Semester") {
      console.log(cur.ariaHidden);
    }
    const next = cur.parentElement;
    if (next === null) {
      return {
        message: () => "Element (or parent) is not aria-hidden",
        pass: false
      };
    }
    cur = next;
  }
  return {
    message: () => "Element (or parent) is aria-hidden",
    pass: true
  };
};

expect.extend({
  toBeUnique,
  toBeHidden
});

declare module "expect" {
  interface AsymmetricMatchers {
    toBeUnique: () => void;
    toBeHidden: () => void;
  }
  interface Matchers<R> {
    toBeUnique: () => R;
    toBeHidden: () => R;
  }
}

export function courseSemestersCheck(semesters: string): void {
  /*
    Available options: FA,WI,SP,SU

    semesters can start with any of the options
    then be followed by a comma and any of the options

    Note: Data such as FA,FA will match but FA,bad_data will not
  */
  // Expect the format to match
  expect(semesters).toMatch(/^(FA|WI|SP|SU)(,(FA|WI|SP|SU))*$/);
}

export function createMockToken(): void {
  updateClient((c) => {
    c.verifyIdToken = async (options: VerifyIdTokenOptions) => {
      const [test, id] = options.idToken.split(":");
      if (test !== "TEST_TOKEN") {
        throw new Error("Invalid Token");
      }
      return new LoginTicket("", {
        sub: id,
        aud: c._clientId ?? "",
        exp: Date.now() + 1000000,
        iat: Date.now() - 100,
        iss: "https://accounts.google.com",
        at_hash: ""
      });
    };
  });
}

export function mockToken(id: string): string {
  return `TEST_TOKEN:${id}`;
}

export function mockUserInfo(id: string): User {
  return {
    cred: mockToken(id),
    info: {
      sub: id,
      picture: "",
      name: id,
      email: `${id}@gmail.com`
    }
  };
}

const db = {
  db: new PromisedDatabase(),
  opened: false
};

export async function setupMockUserDB(): Promise<void> {
  createMockToken();
  setUserDb(db.db);
  if (!db.opened) {
    db.opened = true;
    await db.db.open(":memory:");
    await db.db.run(
      "CREATE TABLE `user` ( `idUser` TEXT PRIMARY KEY, `role` TEXT NOT NULL)"
    );
    await db.db.run(
      "CREATE TABLE `schedule` ( `userID` TEXT, `name` TEXT, `timestamp` DATE DEFAULT (datetime('now','localtime')), `scheduleData` TEXT NOT NULL, PRIMARY KEY(`userID`, `name`))"
    );
  }
}
