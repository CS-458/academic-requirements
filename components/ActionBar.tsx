/*
  Nick Raffel
  This is just a component for the Action Bar on the left Side that houses the save button
*/

import React from "react";
import { Box } from "@mui/material";
import ScheduleUpload from "./ScheduleUploadModal";
import { UserSavedSchedule } from "../entities/four_year_plan";

// Schedule Data and SetAlertData are being passed through here into the Schedule Upload Modal
export default function BoxSx(props: {
  scheduleData: UserSavedSchedule["scheduleData"],
  setAlertData: (msg: string, severity: string) => void
}): any {
  return (
    <div>
      <Box
        sx={{
          height: "100%",
          backgroundColor: "white",
          paddingTop: "2em",
          boxShadow: 9
        }}
      >
        <ScheduleUpload
          scheduleData={props.scheduleData}
          setAlertData={props.setAlertData}
        />
      </Box>
    </div>
  );
}
