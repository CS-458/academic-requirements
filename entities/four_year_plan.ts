import type { TargetType } from "dnd-core";
import { CourseError } from "../components/FourYearPlanPage";

// Defines the major object we are passing between files
export interface MajorType {
  id: number; // database major id
  name: string; // name of the major
}

// Defines the concentration object we are passing between files
export interface ConcentrationType {
  idConcentration: number; // Database concentration id
  name: string; // concentration name
  fourYearPlan: string; // this is a json of a fourYearPlan stored in a string
}

// Defines the course object as it is used in most locations
export interface CourseType {
  credits: number; // number of credits a course has
  name: string; // course name
  number: string; // course number STORED IN DATABASE AS STRING
  semesters: string; // whether it runs in fall or spring (FA or SP) or null (runs both)
  subject: string; // course subject i.e. CS, MATH
  preReq: string; // string containing prerequisite courses in a special format
  category: string; // the category the course falls into
  idCourse: number; // Database course id
  idCategory: number; // Database id for the category
  dragSource: string; // Which element it came from i.e. "Semester 1" or CourseList
  repeatableForCred: boolean; // If a course can be taken twice for credit
}

// Defines the actual draggable course object type
export interface DragCourseType {
  name: string; // see CourseType
  subject: string; // see CourseType
  number: string; // see CourseType
  type: string; // Don't worry about this, all assigned to course
  credits: number; // see CourseType
  semesters: string; // see CourseType
  preReq: string; // see CourseType
  dragSource: string; // see CourseType
  warningYellowColor: number | undefined; // If the course is in multiple semesters
  warningOrangeColor: number | undefined; // If the course is in the wrong semester fall vs spring
  warningRedColor: number | undefined; // If the course has a prereq error
  idCourse: number; // see CourseType
  idCategory: number; // see CourseType
  repeatableForCred: boolean; // see CourseType
}

// Defines the props for the course list component
export interface CourseListType {
  accept: TargetType; // makes it a drop target
  onDrop: (item: any) => void; // function for what to do on drop
  courses: CourseType[]; // list of courses it displays for dragging
}

// Defines a requirement as it comes from the database
export interface RequirementType {
  courseCount: number; // number of courses required to satisfy
  courseReqs: string; // list of courses required to satisfy
  creditCount: number; // number of credits required to satisfy
  idCategory: number; // Database id for requirement
  name: string; // name of the requirement
  parentCategory: number; // Database id for parent requirement category
}

// Defines the saved information for each requirement
export interface RequirementComponentType {
  courseCount: number | null; // number of courses required to satisfy
  courseReqs: string | null; // list of courses required to satisfy
  creditCount: number; // number of credits required to satisfy
  idCategory: number; // Database id for requirement
  name: string; // name of the requirement
  shortName: string; // abbreviated version of the name for display
  parentCategory: number | null; // Database id for parent requirement category
  percentage: number; // Percentage that requirement has been filled
  inheritedCredits: number; // Number of credits a parent req inherits from child req
  coursesTaken: string; // list of courses already used toward requirement
  courseCountTaken: number; // number of courses taken towards requirement
  creditCountTaken: number; // number of credits taken towards requirement
}

// Defines a list of Requirements
export interface RequirementsType {
  [category: number]: RequirementType;
}

// Defines a list of courses in more than one category for requirements
export interface MultipleCategoriesType {
  categories: number[]; // list of category IDs the course falls under
  idString: string; // an identifier for the course i.e. CS-144
}

// define what can be a season
export enum season {
  Fall = "Fall",
  Winter = "Winter",
  Spring = "Spring",
  Summer = "Summer"
}
// define
export enum warning {
  Low = "Low",
  High = "High"
}

export function seasonNum(s: season): number {
  switch (s) {
    case season.Fall:
      return 0;
    case season.Winter:
      return 1;
    case season.Spring:
      return 2;
    case season.Summer:
      return 3;
  }
}

// Defines the properties of a semester component
export interface SemesterProps {
  accept: TargetType; // makes it a drop target, don't worry about it
  onDrop: (item: any) => void; // function for when an item is dropped
  semesterNumber: number; // the number of the semester
  courses: CourseType[]; // a list of courses in the semester
  SemesterCredits: number; // total credits of all courses in semester
  Warning: warning | null; // credit warning (low or high)
  warningPrerequisiteCourses: CourseError[]; // list of courses in with prereq issues
  warningFallvsSpringCourses: CourseError[];
  warningDuplicateCourses: CourseError[]; // list of courses in more than one semester
  year: number; // year number 1,2,3,4,etc.
  season: season; // Season the semester is (Fall, Winter, Spring, Summer)
}

// Defines the properties of a semester object
export interface SemesterType {
  accepts: string[]; // this is just a course constant
  semesterNumber: number; // number of the semester
  courses: CourseType[]; // list of courses in semester
  SemesterCredits: number; // number of credits in the semester
  Warning: warning | null; // credit warning (high or low)
  year: number; // year number 1,2,3,4,etc.
  season: season; // Season the semester is (Fall, Winter, Spring, Summer)
}

export function sortSemester(a: SemesterType, b: SemesterType): number {
  const aNum = seasonNum(a.season) + a.year * 4;
  const bNum = seasonNum(b.season) + b.year * 4;
  return aNum - bNum;
}

//  Defines the properties that should be passed in to the fourYearPlan page
export interface FourYearPlanType {
  PassedCourseList: CourseType[];
  requirements: RequirementComponentType[] | null | undefined;
  requirementsGen: RequirementComponentType[] | null | undefined;
  importData?: {};
}

export interface UserSavedSchedule {
  userID: string;
  name: string;
  timestamp: string;
  scheduleData: {
    Major: number;
    Concentration: number;
    "Completed Courses": string[];
    schedule: ScheduleData;
  };
}

export interface movedCourse {
  movedTo: number,
  movedFrom: number,
  course: number
}
export type ScheduleData = Array<{
  year: number;
  seasons: Array<{ season: season; classes: string[] }>;
}>;
