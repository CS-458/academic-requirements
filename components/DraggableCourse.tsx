import type { FC } from "react";
import React, { memo } from "react";
import { useDrag } from "react-dnd";
import clsx from "clsx";
import { DragCourseType } from "../entities/four_year_plan";

export const Course: FC<DragCourseType> = memo(function Course({
  name,
  subject,
  number,
  type,
  credits,
  semesters,
  preReq,
  dragSource,
  warningYellowColor,
  warningOrangeColor,
  warningRedColor,
  idCourse,
  idCategory,
  repeatableForCred
}) {
  // defines the drag action
  const [{ opacity }, drag] = useDrag(
    () => ({
      type,
      item: {
        name,
        subject,
        number,
        type,
        credits,
        semesters,
        preReq,
        dragSource,
        warningYellowColor,
        warningOrangeColor,
        warningRedColor,
        idCourse,
        idCategory,
        repeatableForCred
      },
      collect: (monitor: any) => ({
        opacity: monitor.isDragging() !== false ? 0.8 : 1
      })
    }),
    [idCourse, type, dragSource] // what is collected by the semester and course list when you drop it
  );

  // Gets the URL to the UW Stout Bulletin for the given Course
  function getURL(subject: string, number: string): string {
    const URL =
      "https://bulletin.uwstout.edu/content.php?filter%5B27%5D=" +
      subject +
      "&filter%5B29%5D=" +
      number +
      "&filter%5Bcourse_type%5D=-1&filter%5Bkeyword%5D=&filter%5B32%5D=1&filter%5Bcpage%5D=1&cur_cat_oid=21&expand=&navoid=544&search_database=Filter#acalog_template_course_filter";
    return URL;
  }

  return (
    <div
      ref={drag}
      style={{ opacity }}
      data-testid="course"
      className={clsx(
        "CourseText",
        warningYellowColor !== undefined &&
        warningRedColor === undefined &&
        warningOrangeColor === undefined &&
        "CourseWarningYellow",
        warningOrangeColor !== undefined &&
        warningRedColor === undefined &&
        "CourseWarningOrange",
        warningRedColor !== undefined && "CourseWarningRed"
      )}
      key={`course-${idCourse}`}
    >
      {/* {isDropped ? <s>{name}</s> : name}  */}
      {subject}-{number}
      <br />
      {name}
      <br />
      credits: {credits}
      <br />
      <a href={getURL(subject, `${number}`)} target="_blank">
        Description
      </a>
    </div>
  );
});
