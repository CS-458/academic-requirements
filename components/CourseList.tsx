import type { CSSProperties, FC } from "react";
import React, { memo } from "react";
import { useDrop } from "react-dnd";
import { ItemTypes } from "../entities/Constants";
import { Course } from "./DraggableCourse";

// Styling for the course list
const style: CSSProperties = {
  height: "30rem",
  width: "90%",
  marginRight: ".5rem",
  marginBottom: ".5rem",
  color: "white",
  padding: "1rem",
  textAlign: "center",
  fontSize: "1rem",
  lineHeight: "normal",
  float: "left",
  background: "#004990",
  borderRadius: ".5rem",
  overflow: "auto"
};

export interface CourseListProps {
  accept: Course;
  onDrop: (item: any) => void;
  courses: Course[];
}

export const CourseList: FC<CourseListProps> = memo(function CourseList({
  accept,
  onDrop,
  courses
}) {
  // defines the drop
  const [{ isOver }, drop] = useDrop({
    accept,
    drop: onDrop,
    collect: (monitor) => ({
      isOver: monitor.isOver()
    })
  });

  // changes the background color on hover over course list
  const isActive = isOver;
  let backgroundColor = "#004990";
  if (isActive) {
    backgroundColor = "darkgreen";
  }

  return (
    <div ref={drop} style={{ ...style, backgroundColor }}>
      {isActive ? "Release to drop" : "CourseList"}
      {courses.map(
        ({ name, subject, number, semesters, credits, preReq }, index) => (
          <Course
            name={name}
            subject={subject}
            number={number}
            semesters={semesters}
            credits={credits}
            type={ItemTypes.COURSE}
            key={index}
            dragSource={"CourseList"}
            preReq={preReq}
          />
        )
      )}
    </div>
  );
});
