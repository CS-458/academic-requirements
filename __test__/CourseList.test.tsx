import "@testing-library/jest-dom";
import { jest } from "@jest/globals";
import { setupUser, render } from "./util";
import { CourseList } from "../components/CourseList";
import { Course } from "../components/DraggableCourse";
import { CourseType, DragCourseType } from "../entities/four_year_plan";
import { ItemTypes } from "../entities/Constants";
import { fireEvent, screen } from "@testing-library/react";

const testCourseData: CourseType[] = [
  {
    subject: "TEST-SUB-1",
    number: "100HON",
    name: "Course 1",
    semesters: "",
    credits: 1,
    preReq: "",
    idCourse: 294,
    repeatableForCred: false,
    category: "TEST-CATEGORY-1",
    idCategory: 20,
    dragSource: ""
  },
  {
    subject: "TEST-SUB-2",
    number: "200",
    name: "Course 2",
    semesters: "",
    credits: 2,
    preReq: "",
    idCourse: 295,
    repeatableForCred: false,
    category: "TEST-CATEGORY-2",
    idCategory: 20,
    dragSource: ""
  },
  {
    subject: "TEST-SUB-2",
    number: "200",
    name: "Course 2",
    semesters: "",
    credits: 2,
    preReq: "",
    idCourse: 295,
    repeatableForCred: false,
    category: "TEST-CATEGORY-3",
    idCategory: 20,
    dragSource: ""
  },
  {
    subject: "TEST-SUB-3",
    number: "300",
    name: "Course 3",
    semesters: "",
    credits: 3,
    preReq: "",
    idCourse: 296,
    repeatableForCred: false,
    category: "TEST-CATEGORY-3",
    idCategory: 20,
    dragSource: ""
  }
];

test("CourseList Renders", () => {
  const index = render(
    <CourseList
      accept={[ItemTypes.COURSE]}
      onDrop={jest.fn()}
      courses={testCourseData}
    />
  );
  expect(index.baseElement).toMatchSnapshot();
});

test("CourseList onDrop is called on drop", () => {
  const onDropMock = jest.fn();
  const index = render(
    <CourseList
      accept={[ItemTypes.COURSE]}
      onDrop={onDropMock}
      courses={testCourseData}
    />
  );
  expect(index.baseElement).toMatchSnapshot();
  // The outermost dive within the CourseList component
  const root = screen.getByTestId("test-root-element").children[0];

  testCourseData.forEach((c, index) => {
    fireEvent.dragStart(root.children[index]);
    fireEvent.dragEnter(root);
    fireEvent.dragOver(root);
    expect(root).toHaveStyle("background-color: darkgreen");
    fireEvent.drop(root);
    console.log(onDropMock.mock.calls);
    expect(onDropMock).toHaveBeenCalled();
    // expect(onDropMock).toHaveBeenCalledWith(c as unknown as DragCourseType);
  });
});
