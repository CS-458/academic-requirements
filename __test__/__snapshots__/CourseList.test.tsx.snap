// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CourseList Renders 1`] = `
<body>
  <div>
    <div
      data-testid="test-root-element"
    >
      <div
        style="height: 30rem; width: 100%; margin-right: .5rem; margin-bottom: .5rem; color: white; padding: 1rem; text-align: center; font-size: 1rem; line-height: normal; float: left; background: rgb(0, 73, 144); border-radius: .5rem; overflow: auto; background-color: rgb(0, 73, 144);"
      >
        <div
          class="CourseText"
          data-testid="course"
          draggable="true"
          style="opacity: 1;"
        >
          TEST-SUB-1
          -
          100HON
          <br />
          Course 1
          <br />
          credits: 
          1
          <br />
          <a
            href="https://bulletin.uwstout.edu/content.php?filter%5B27%5D=TEST-SUB-1&filter%5B29%5D=100HON&filter%5Bcourse_type%5D=-1&filter%5Bkeyword%5D=&filter%5B32%5D=1&filter%5Bcpage%5D=1&cur_cat_oid=21&expand=&navoid=544&search_database=Filter#acalog_template_course_filter"
            target="_blank"
          >
            Description
          </a>
        </div>
        <div
          class="CourseText"
          data-testid="course"
          draggable="true"
          style="opacity: 1;"
        >
          TEST-SUB-2
          -
          200
          <br />
          Course 2
          <br />
          credits: 
          2
          <br />
          <a
            href="https://bulletin.uwstout.edu/content.php?filter%5B27%5D=TEST-SUB-2&filter%5B29%5D=200&filter%5Bcourse_type%5D=-1&filter%5Bkeyword%5D=&filter%5B32%5D=1&filter%5Bcpage%5D=1&cur_cat_oid=21&expand=&navoid=544&search_database=Filter#acalog_template_course_filter"
            target="_blank"
          >
            Description
          </a>
        </div>
        <div
          class="CourseText"
          data-testid="course"
          draggable="true"
          style="opacity: 1;"
        >
          TEST-SUB-2
          -
          200
          <br />
          Course 2
          <br />
          credits: 
          2
          <br />
          <a
            href="https://bulletin.uwstout.edu/content.php?filter%5B27%5D=TEST-SUB-2&filter%5B29%5D=200&filter%5Bcourse_type%5D=-1&filter%5Bkeyword%5D=&filter%5B32%5D=1&filter%5Bcpage%5D=1&cur_cat_oid=21&expand=&navoid=544&search_database=Filter#acalog_template_course_filter"
            target="_blank"
          >
            Description
          </a>
        </div>
        <div
          class="CourseText"
          data-testid="course"
          draggable="true"
          style="opacity: 1;"
        >
          TEST-SUB-3
          -
          300
          <br />
          Course 3
          <br />
          credits: 
          3
          <br />
          <a
            href="https://bulletin.uwstout.edu/content.php?filter%5B27%5D=TEST-SUB-3&filter%5B29%5D=300&filter%5Bcourse_type%5D=-1&filter%5Bkeyword%5D=&filter%5B32%5D=1&filter%5Bcpage%5D=1&cur_cat_oid=21&expand=&navoid=544&search_database=Filter#acalog_template_course_filter"
            target="_blank"
          >
            Description
          </a>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`CourseList onDrop is called on drop 1`] = `
<body>
  <div>
    <div
      data-testid="test-root-element"
    >
      <div
        style="height: 30rem; width: 100%; margin-right: .5rem; margin-bottom: .5rem; color: white; padding: 1rem; text-align: center; font-size: 1rem; line-height: normal; float: left; background: rgb(0, 73, 144); border-radius: .5rem; overflow: auto; background-color: rgb(0, 73, 144);"
      >
        <div
          class="CourseText"
          data-testid="course"
          draggable="true"
          style="opacity: 1;"
        >
          TEST-SUB-1
          -
          100HON
          <br />
          Course 1
          <br />
          credits: 
          1
          <br />
          <a
            href="https://bulletin.uwstout.edu/content.php?filter%5B27%5D=TEST-SUB-1&filter%5B29%5D=100HON&filter%5Bcourse_type%5D=-1&filter%5Bkeyword%5D=&filter%5B32%5D=1&filter%5Bcpage%5D=1&cur_cat_oid=21&expand=&navoid=544&search_database=Filter#acalog_template_course_filter"
            target="_blank"
          >
            Description
          </a>
        </div>
        <div
          class="CourseText"
          data-testid="course"
          draggable="true"
          style="opacity: 1;"
        >
          TEST-SUB-2
          -
          200
          <br />
          Course 2
          <br />
          credits: 
          2
          <br />
          <a
            href="https://bulletin.uwstout.edu/content.php?filter%5B27%5D=TEST-SUB-2&filter%5B29%5D=200&filter%5Bcourse_type%5D=-1&filter%5Bkeyword%5D=&filter%5B32%5D=1&filter%5Bcpage%5D=1&cur_cat_oid=21&expand=&navoid=544&search_database=Filter#acalog_template_course_filter"
            target="_blank"
          >
            Description
          </a>
        </div>
        <div
          class="CourseText"
          data-testid="course"
          draggable="true"
          style="opacity: 1;"
        >
          TEST-SUB-2
          -
          200
          <br />
          Course 2
          <br />
          credits: 
          2
          <br />
          <a
            href="https://bulletin.uwstout.edu/content.php?filter%5B27%5D=TEST-SUB-2&filter%5B29%5D=200&filter%5Bcourse_type%5D=-1&filter%5Bkeyword%5D=&filter%5B32%5D=1&filter%5Bcpage%5D=1&cur_cat_oid=21&expand=&navoid=544&search_database=Filter#acalog_template_course_filter"
            target="_blank"
          >
            Description
          </a>
        </div>
        <div
          class="CourseText"
          data-testid="course"
          draggable="true"
          style="opacity: 1;"
        >
          TEST-SUB-3
          -
          300
          <br />
          Course 3
          <br />
          credits: 
          3
          <br />
          <a
            href="https://bulletin.uwstout.edu/content.php?filter%5B27%5D=TEST-SUB-3&filter%5B29%5D=300&filter%5Bcourse_type%5D=-1&filter%5Bkeyword%5D=&filter%5B32%5D=1&filter%5Bcpage%5D=1&cur_cat_oid=21&expand=&navoid=544&search_database=Filter#acalog_template_course_filter"
            target="_blank"
          >
            Description
          </a>
        </div>
      </div>
    </div>
  </div>
</body>
`;
